### ISSUE TITLE
batch_naive neighbor matrix and dftd3 are not run-to-run deterministic on CUDA

### ISSUE THREAD

**Issue by isayev** [NVIDIA/nvalchemi-toolkit-ops#135]
## Summary

Two CUDA ops produce non-reproducible results for repeated identical inputs:

1. `neighbor_list(..., method="batch_naive")` returns neighbor matrices whose per-row *content* is stable but whose within-row *order* varies run to run.
2. `dftd3(...)` returns energies that vary at the accumulation-rounding level even when called repeatedly with a **frozen** neighbor matrix.

Each effect alone is ~1e-9 Hartree — physically negligible — but downstream consumers that feed the forces to iterative geometry optimizers see chaotic trajectory divergence from byte-identical inputs (batched optimization campaigns are not repeatable; details in isayevlab/aimnetcentral#93).

## Reproducer 1: neighbor order instability

```python
import torch
from nvalchemiops.torch.neighbors import neighbor_list

DEV = "cuda"
torch.manual_seed(0)
B, N = 8, 24
coord = (torch.rand(N, 3, device=DEV) * 6.0).unsqueeze(0).expand(B, -1, -1).contiguous().flatten(0, 1)
batch_idx = torch.arange(B, device=DEV, dtype=torch.int32).repeat_interleave(N)

def build():
    nbmat, _ = neighbor_list(positions=coord, cutoff=15.0, batch_idx=batch_idx,
                             max_neighbors=64, half_fill=False, fill_value=B*N, method="batch_naive")
    return nbmat

n1, n2 = build(), build()
print("bitwise identical:", torch.equal(n1, n2))                          # False
print("sorted rows identical:", torch.equal(n1.sort(1).values, n2.sort(1).values))  # True
```

## Reproducer 2: dftd3 varies with a frozen neighbor matrix

Calling `dftd3` 10 times with the *same* positions, numbers, and neighbor matrix (built once with reproducer 1's inputs, standard wB97M D3(BJ) parameters, `batch_idx` as above) gives `max|dE|` up to **3.7e-09 Hartree** between calls, so the variance is not only neighbor ordering — the kernel's own accumulation order varies.

## Environment

- nvalchemi-toolkit-ops 0.4.0 (also reproduced on 0.3.1; on 0.3.1 batch sizes <= 2 were bitwise stable, on 0.4.0 they no longer are)
- warp-lang 1.15.0, torch 2.9.1+cu128
- NVIDIA L40S, CUDA 12.9, driver 13.2

## Ask

Is run-to-run determinism a supported guarantee for these ops? If the atomics are a deliberate performance trade (understood!), an opt-in deterministic reduction mode — or a documented statement that batched results are reproducible only to accumulation rounding — would let downstream packages route reproducibility-sensitive workloads accordingly.

**Comment by zubatyuk** [comment]
The current neighbor list contract doesn't enforce a specific neighbor order. The non-determinism you're seeing is expected because the implementation uses atomic operations. You are correct that sorting the rows of the neighbor matrix is the right approach to make it stable.

Could you share an example of a practical scenario where the numerical accuracy variations in the DFT-D3 implementation pose a problem?
